New notebook - Interactive multi-wavelength visualizations of named sources (#672)
fornax-navo/fornax-demo-notebooks: multiarchive_visualization/multiwavelength_image.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.19.1
kernelspec:
  name: py-multiwavelength_image
  display_name: py-multiwavelength_image
  language: python
authors:
  - name: David J Turner
---

# Create interactive multi-wavelength images of astronomical sources

## Learning goals

By the end of this tutorial, you will be able to:
- Find Chandra, Hubble, Spitzer, and Swift observations of a named source.
- Reproject images from different missions to a common coordinate grid.
- Create interactive visualizations of individual reprojected images.

> - Interactively explore how 'pretty' RGB images can be made from multi-wavelength observations.

Reviewer: ```suggestion
- Interactively explore how visually interesting RGB images can be made from multi-wavelength observations.
```

## Introduction

This notebook is intended as a gentle introduction to the sort of things you can do 
on the Fornax Science Console, while also providing a demonstration of how you might
begin to make visually appealing multi-wavelength images of astronomical sources.
(Though the author makes no claims as to having a _good_ sense of aesthetics – you're on 
your own there).

We'll be searching the NASA astrophysics archives for observations of named 
sources, with a set of suggested objects to try out.
We will also be setting up interactive visualizations where you can adjust how different wavelengths are combined and
see what makes the nicest looking image.

Note that we will only be using pre-generated images, rather than making any new
ones of our own from the raw observation data.

Throughout this demonstration we make use of the Python module Astroquery to search and 

> acquire data from the HEASARC, MAST, and IRSA archives.

Reviewer 2: Define all three of these acronyms here. You can remove the definitions that appear lower in the notebook.

Might also be good to add links here.

The Matplotlib and ipywidgets libraries are used to create interactive figures that are performant even with
the thousands-of-pixels-per-side images we will be dealing with.

### Input
- The name of the target source; *we provide a few suggestions in the first section of this notebook.*

### Output

- Interactive multi-wavelength images of the target source.

### Runtime

> As of 21st April 2026, this notebook takes approximately 8 minutes to run to completion on Fornax using the small server with 8GB RAM/ 2 CPU, with the default settings.

Reviewer: update date and timing when you do edits.  Also, I am not sure there are any settings other than server size, so either clarify "default settings" or remove.

This demonstration acquires data from remote services, and as such the runtime can vary depending on the state of those services and the speed of your internet connection (if running locally).

## Imports

```{code-cell} python
# Uncomment the next line to install dependencies if needed.
# %pip install -r requirements_multiwavelength_image.txt
```

```{code-cell} python
import sys

from astropy.coordinates import SkyCoord

> from astropy.io import fits
> from astropy.time import Time

Reviewer 2: unused

```suggestion
```

from astropy.units import Quantity
from astroquery.heasarc import Heasarc
from astroquery.ipac.irsa import Irsa
from astroquery.mast import Observations

from IPython.display import display

from astropy import conf
conf.max_lines = 6

import numpy as np

# Add local code directory to path
sys.path.append('code_src/')

# Import our custom functions
from archive_queries import (vetted_source_check, load_chandra_image, 
                             load_spitzer_image, load_hubble_image, load_swift_image)
from image_processing import get_pixel_scale, reproject_to_common_grid
from plotting import InteractiveRGBPanel, InteractiveMultiPanel, get_vis_presets
```

+++

## 1. Choosing the object we want to visualize

We start by choosing the source for which we want to create a multi-wavelength image; as we are 
using the `SkyCoord.from_name(...)` method to initialize the SkyCoord object, you may any name that 
can be identified by the CDS name resolver service. Alternatively, you could manually set up 
a SkyCoord from the RA-Dec (or other coordinate system) of your source of interest.

> To get started, we provide a few suggestions to try out:

Reviewer 2: I think these are what you call "vetted" and provide extra functionality for? If so, add description here of what will be different if not using one of these suggested objects.

- ***The Crab Nebula*** **[Crab; default]** – One of the most observed sources in the Milky Way, a favorite for calibrating space observatories, and famously visually striking.
- ***Messier 61*** – Also known as the 'Swelling Spiral Galaxy', M61 is one of the largest members of the Virgo galaxy cluster and has played host to 8 observed supernovae since 1926 (a considerable number). 
- ***ZwCl 0024+1652*** - A galaxy cluster that demonstrates significant structure in its dark matter distribution.
- ***Kepler's Supernova*** **[SN 1604]** – The remnant of the most recent supernova observed with the naked eye (in 1604).
- ***Abell 370*** - A galaxy cluster known for several prominent strong-lensing arc features.

> - ***NGC 2595*** - A large elliptical galaxy, part of a galaxy group, discovered by William Herschel in 1787.

Reviewer: NGC2595 is a barred spiral.

- ***IC 5201*** - A bar spiral galaxy with areas of significant star formation.

```{code-cell} python
# Define the target
SOURCE_NAME = "Crab"

# Other source suggestions - uncomment them (and comment out the others) to try them out.
# SOURCE_NAME = "M61"
# SOURCE_NAME = "ZwCl0024+1652"
# SOURCE_NAME = "SN1604"
# SOURCE_NAME = "A370"
# SOURCE_NAME = "NGC2595"
# SOURCE_NAME = "IC5201"

# Resolve coordinates from name
SOURCE_COORD = SkyCoord.from_name(SOURCE_NAME)

print(f"{SOURCE_NAME} Coordinate:".upper())
print(SOURCE_COORD.to_string())
```

+++

## 2. Searching NASA's astrophysical archives for images of our source

Our goal is to be able to compare and combine images of the same source, taken in 
different wavelength bands.

> Some space telescopes (Hubble, for instance) can take 

Reviewer: I believe most (if not all) space telescopes have multi-wavelength observations.  Either way, I don't think this is a necessary statement for a target audience of astronomers.  Suggest removing it.

observations in several, usually adjacent, wavelength bands.
However, for this 
notebook, and to explore a wider range of physical processes highlighted by different
wavelengths of light, we will search for observations taken by four **different** 
observatories:
- Chandra [X-ray photons]
- Swift [UV photons]

> - Spitzer Infrared Telescope [Infrared photons]

Reviewer: ```suggestion
- Spitzer Space Telescope [Infrared photons]
```

Reviewer: actual name.

- Hubble Space Telescope [Optical photons]

### 2.1 Query HEASARC for Chandra X-ray observations

Chandra has the highest angular resolution of any X-ray telescope, which is of particular
importance given that we want to make nice visualizations of our target.
As high-energy photons are much harder to focus than their lower-energy cousins, even 
Chandra's spatial resolution is significantly worse than that of Hubble.

The High Energy Astrophysics Science Archive Research Center (HEASARC) provides 
access to the full public set of Chandra observations.
We'll use the `Heasarc` 
object imported from the `astroquery.heasarc` module to search for observations of 
current target.

> First, though, we can check if the current target source has been pre-vetted to identify

Reviewer: clarify what "pre-vetted" means?  here, and in the function docstring.

```{code-cell} python
chandra_obs_id = vetted_source_check(SOURCE_NAME, "Chandra")
chandra_obs_id
```

Now, we need to search for observations of our current target.
Note though that if 
there is a pre-vetted observation for the source, we won't consider any other 
observations that might be relevant to the source.

We set a search radius of $3^{\prime}$, which should help exclude any observations 
where the source is not on or near the CCD chip positioned at the aimpoint.
You could
experiment with adjusting this radius if you aren't finding observations of your 
source:

```{code-cell} python
CHANDRA_SEARCH_RAD = Quantity(3, 'arcmin')
```

Our search then begins by defining filters, to specify which Chandra instruments we might
want to consider using for our visualizations:
- **ACIS-S/I** - Chandra's CCD spectro-imagers, these are much more commonly used than HRC-I/S (the other Chandra instruments).
- **No grating** – If a grating was deployed, then the resulting 'image' would actually be a dispersed spectrum trace, rather than an image of the source.

Note that pre-vetted ObsID is added to the filters, if available.

> Following the definition of our filters, we pass the coordinate of the source (specified 

Reviewer: this sentence is missing a phrase about *where* the coords, catalog name, filters, and column names arguments get passed.  

suggest something along the lines of "we pass to the the astroquery function Heasarc.query_region()...."

we have on hand.

```{code-cell} python
search_filt = {"detector": ["ACIS-S", "ACIS-I"], 
               "grating": "NONE"}

# If a pre-vetted observation exists, we'll add an extra filter to only select that
#  one observation.              
search_filt.update({} if chandra_obs_id is None else {"obsid": chandra_obs_id})

all_chandra_obs = Heasarc.query_region(SOURCE_COORD, 
                                       catalog='chanmaster', 
                                       column_filters=search_filt, 
                                       columns='*', 
                                       radius=CHANDRA_SEARCH_RAD)
                                       
all_chandra_obs.sort('exposure', reverse=True)

all_chandra_obs
```

If there are any observations of the target source, we will select the first entry, 
with the longest exposure time; you could also experiment with choosing other 
observations by filtering the `all_chandra_obs` table yourself.

From there, we use another `Heasarc` function to fetch the 'data link' for our chosen
observation, which will tell us where we can fetch the data files from:

```{code-cell} python

> if len(all_chandra_obs) > 0:
>     sel_chandra_datalink = Heasarc.locate_data(all_chandra_obs[0])
> else:
>     sel_chandra_datalink = None

Reviewer 2: If making `load_chandra_image()` accept only `str` or `None`

```suggestion
if len(all_chandra_obs) > 0:
    sel_chandra_datalink = Heasarc.locate_data(all_chandra_obs[0])
    chosen_chandra_uri = str(sel_chandra_datalink['aws'][0])
else:
    chosen_chandra_uri = None
```

```

Now we use a convenience function (see the `code_src/archive_queries.py` file for the 
definition) to use the data link to fetch and load the Chandra image:

```{code-cell} python

> chandra_hdu = load_chandra_image(sel_chandra_datalink, preproc_cent_hi_res=True)

Reviewer 2: If renaming the variable

```suggestion
chandra_hdu = load_chandra_image(chosen_chandra_uri, preproc_cent_hi_res=True)
```

```

### 2.2 Query IRSA for infrared data

For infrared data, we turn to the Spitzer space telescope.
Spitzer is one of NASA's 
'great observatories' (alongside Chandra, Hubble, and Compton) and provides infrared 
imaging (and spectroscopy, but that isn't relevant to this demonstration) in the 
mid/far-IR bands.

> Unlike the Chandra and Hubble great observatories that we're also using in this 
> tutorial, Spitzer is no longer active, but its data are all maintained and served by 
> the NASA/IPAC Infrared Science Archive (IRSA). 

Reviewer: ```suggestion
Spitzer data are all maintained and served by 
the NASA/IPAC Infrared Science Archive (IRSA). 
```

operational status is irrelevant to this tutorial.

```

For simplicity, we're only going to search for images taken by one of Spitzer's 
instruments – the Infrared Array Camera (IRAC). We pass the source coordinate and
search radius, specify that we want image products, and particularly those that have
a calibration level of **2** (science ready calibrated image mosaics) or **3** (enhanced 
data products often made up of data from multiple visits):

```{code-cell} python
all_spitzer_ims = Irsa.query_sia(pos=(SOURCE_COORD, SPITZER_SEARCH_RAD), 
                                 facility="Spitzer Space Telescope", 
                                 data_type="image", 
                                 instrument='IRAC', 
                                 res_format='image/fits', 
                                 calib_level=[2, 3])

# These columns are removed to make the table look nicer when shown in the notebook
del all_spitzer_ims['s_region']
del all_spitzer_ims['proposal_title']

# If a pre-vetted observation exists for the target, we'll filter out all 
#  products related to other observations.
if spitzer_obs_id is not None:
    all_spitzer_ims = all_spitzer_ims[all_spitzer_ims['obs_id'] == spitzer_obs_id]

> # Don't want any calibration products or any such, just science products

Reviewer: ```suggestion
# Don't want any calibration products, just science products
```

"or any such" is vague and therefore not helpful to a user.

    filt_spitzer_ims = filt_spitzer_ims[filt_spitzer_ims['s_resolution'] == filt_spitzer_ims['s_resolution'].min()]
    
    # Filter to mean mosaics (exclude short HDR exposures and median mosaics) - if the
    #  products aren't mosaics at all then this has no effect
    not_short_median_filt = (
        (~np.char.find(filt_spitzer_ims['access_url'].data.astype(str), 'short') > -1) &
        (~np.char.find(filt_spitzer_ims['access_url'].data.astype(str), 'median') > -1)
    )
    filt_spitzer_ims = filt_spitzer_ims[not_short_median_filt]
    
    # Calculate distance of pointing from source
    cur_sep = SkyCoord(filt_spitzer_ims['s_ra'].value, 
                       filt_spitzer_ims['s_dec'].value, 
                       unit='deg').separation(SOURCE_COORD)
    filt_spitzer_ims['dist_to_point'] = cur_sep
    
    # Also sort by the distance from the target (closest at the top of the table)
    filt_spitzer_ims.sort('dist_to_point')
else:
    filt_spitzer_ims = None

# Show the table of products
filt_spitzer_ims
```

We use another convenience function (see the `code_src/archive_queries.py` file for the 
definition) to load the Spitzer image:

```{code-cell} python

> sel_spitzer_im = None if filt_spitzer_ims is None else filt_spitzer_ims[0]
> spitzer_hdu = load_spitzer_image(sel_spitzer_im)

Reviewer 2: If making `load_spitzer_image()` accept only `str` or `None`

```suggestion
spitzer_im_cloud = None if filt_spitzer_ims is None else filt_spitzer_ims[0]['cloud_access']
spitzer_hdu = load_spitzer_image(spitzer_im_cloud)
```

also have disparate spatial resolutions, driven by the available technology and the instrument's
scientific purpose.

As such, we can't just plot the images directly over the top of each other and expect 
everything to line up properly.
Of course, we can't expect that every feature we see in
one image will be present in the others.
The whole point of observing in different
wavelengths is to understand _different_ astrophysical processes going on in a source.

Here we use the image with the highest spatial resolution (almost certain to be the 
Hubble image if we managed to find one for the current source) as the reference 
coordinate grid.
We'll refer to it as the 'donor' image.

That means that all other images will be aligned to the 'donor', and will have to be 
upscaled to match the donor's higher spatial resolution.

:::{note}
We would not necessarily recommend this interpolating and upscaling for images that 
are going to be used for scientific analyses, but given we're just making nice 
visualizations, it is fine to do it here.
:::

### 3.1 Compare pixel scales

```{code-cell} python
# Dictionary mapping missions to their relevant HDU extensions

> mission_hdus = {

Reviewer 2: Add a markdown cell with a little narrative about what this code cell does.

    'Chandra': chandra_hdu['PRIMARY'] if chandra_hdu else None,
    'Hubble': hubble_hdu['SCI'] if hubble_hdu else None,
    'Swift': swift_hdu[1] if swift_hdu else None,
    'Spitzer': spitzer_hdu['PRIMARY'] if spitzer_hdu else None
}

# A check for all HDUs being None
if all([v is None for v in mission_hdus.values()]):
    raise ValueError(f"No data are available for {SOURCE_NAME}.")

pixel_scales = {miss: get_pixel_scale(miss_hdu) for miss, miss_hdu in mission_hdus.items() if miss_hdu is not None}
finest_pix_miss = min(pixel_scales, key=pixel_scales.get)

> 

Reviewer: ```suggestion
Print out a table of pixel scales for the available data.
```

print("-" * 50)
print(f"{finest_pix_miss:<10}: {pixel_scales[finest_pix_miss].to('arcsec/pix'):.3f} [x1.00]")
tab_strs = {}
for cur_miss, cur_scale in pixel_scales.items():
    if cur_miss == finest_pix_miss:
        continue  
    coarse_factor = (cur_scale / pixel_scales[finest_pix_miss]).value.round(1)
    tab_strs[coarse_factor] = f"{cur_miss:<10}: {cur_scale.to('arcsec/pix'):.3f} [x{coarse_factor}]\n"
    
print("".join(dict(sorted(tab_strs.items())).values()))
print("-" * 50)
```

+++

### 3.2 Reprojecting images to a common coordinate grid

We have to reproject the images to a common coordinate grid, giving them a matching 
spatial resolution for when we make visualizations. 

If we were trying to do science with these images directly, we would likely not 
reproject them.

Here we fetch out the highest resolution image's FITS header, from which we can fetch 
the information that will act as the common coordinate grid for all images.
Note though 
that we will not reproject the donor image:
```{code-cell} python
donor_wcs_hdr = mission_hdus[finest_pix_miss].header.copy()
```

```{code-cell} python

> reproj_data_cov = {mn: {"data": (rp := reproject_to_common_grid((cur_hdu.data, cur_hdu.header), donor_wcs_hdr))[0], "cov": ((rp[1] > 0).sum() / rp[1].size)} for mn, cur_hdu in mission_hdus.items() if cur_hdu is not None and mn != finest_pix_miss}

Reviewer 2: ```suggestion
reproj_data_cov = {
    mn: {
        "data": (rp := reproject_to_common_grid((cur_hdu.data, cur_hdu.header), donor_wcs_hdr))[0],
        "cov": ((rp[1] > 0).sum() / rp[1].size),
    }
    for mn, cur_hdu in mission_hdus.items()
    if cur_hdu is not None and mn != finest_pix_miss
}
```

for cur_miss, rp_info in reproj_data_cov.items():
    if rp_info['data'] is not None:
        print(f"{cur_miss}: {rp_info['data'].shape}, coverage = {rp_info['cov'] * 100:.1f}%")
        
        # While we're here, we also remove the original data for the current 
        #  mission from memory, as we have the reprojected data now.
        del mission_hdus[cur_miss].data
        
reproj_data_cov[finest_pix_miss] = {'data': mission_hdus[finest_pix_miss].data, 'cov': 1.0}
reproj_data_cov = {mn: reproj_data_cov[mn] for mn in mission_hdus if mission_hdus[mn] is not None}
```

> 

Reviewer: ```suggestion
# All needed arrays now live in reproj_data_cov as in-memory copies, so we can
  #  release the open S3/file handles from the four load_* calls.
  for hdu in (chandra_hdu, hubble_hdu, spitzer_hdu, swift_hdu):
      if hdu is not None:
          hdu.close()
```

+++

## 4. Visualize individual wavelength images

```{code-cell} python

> sep_reproj_im_cmaps = {

Reviewer 2: Add a markdown cell with a little bit of narrative about what's being visualized and what the interactive controls are.

    'Hubble': 'Greens',
    'Swift': 'Blues',
    'Chandra': 'Purples',
    'Spitzer': 'YlOrBr'
}
```

```{code-cell} python
sep_ims = InteractiveMultiPanel({mn: res['data'] for mn, res in reproj_data_cov.items() if res['cov'] > 0.1}, sep_reproj_im_cmaps)

> sep_ims.view()

Reviewer 2: Is it possible to add a title with the name of the object?

```

+++

> ## 5. Interactive multi-wavelength image

Reviewer 2: Good explanations in this section.

Finally, we combine three images in different wavelengths into a single RGB composite image.
We use:

> - Red channel: Spitzer infrared (cool dust and molecular gas)

Reviewer: ```suggestion
- Red channel: Spitzer infrared ("warm dust, PAH emission, and evolved stars.")
```

- Green channel: Hubble optical (stellar emission and ionized gas)
- Blue channel: Chandra X-ray (hot plasma and high-energy astrophysical processes)

> The interactive controls allow you to adjust:

Reviewer: for me this cell first loads as just a single colored row, but once I start messing with the percentiles - the full image renders.  No idea how to fix that. running on fornax.
